### DISCUSSION TITLE
Report - (4th Sem.) Week 4 📢

### DISCUSSION THREAD

**Discussion by giomskii** [SE-TINF22B6/DHBWhub#595] [Weekly Report]
Greetings and welcome to our twelth progress report! We're thrilled to update you on the latest news and developments since last week! 🥳

### Highlights of our test plan 🧪 
In this part of our weekly blog, we will summarize our current status regarding the testing of our DHBWhub application. If you are even more interested in this topic, you can read our full test report right [here](https://github.com/SE-TINF22B6/DHBWhub/blob/master/docs/TEST_REPORT.md)!  
#### Frontend Testing 
In the frontend, these points are crucial when it comes to testing:
- Manually checking, whether all components render correctly and also on all devices (Capturing differences between smartphone and desktop view).
- Unit and integration tests to determine, if the components react the way they should when an user-caused event gets triggered as well as the correct rerendering of objects.
- Checking, if data is formatted in the way it should.  
 
To fullfill these tests on the frontend, we are planning to use the testing framework 'React testing library'.
  
#### Backend Testing
Our backend consists of three layers (Controller, Service and Repository), where a component in one layer always relies on a component in the layer below. Thus, a mix of unit and integration tests are required to check, whether the given input data returns the excepted data in the correct data format with all the relevant and needed information. We are using 'JUnit5' and 'Mockito' for writing these tests as well as the building tool 'Maven' which automatically runs all the tests which are located in our test-folders. This also happens in our CI/CD-pipeline. Therefore, we can be sure that our current branch or the newly created feature branch doesn't destroy the output of our previous tests!

#### Database Testing
Besides the manual checking, whether the refreshed or newly created views are equivalent to the normal SQL-requests, we also plan to test in the future, how the database handles multiple table operations in a short period of time. 

#### New components!
We have implemented some nice additions to our frontend which you can see in the picture below!  
  
![image](https://github.com/SE-TINF22B6/DHBWhub/assets/111356450/16087d39-3b4c-49c8-a43e-20b7f773e7a3)
  
Our first addition is the "SavedPost"-component, which displays all the posts, that a user has saved/bookmarked for later. It currently depicts the title of the respective post. Secondly, posts are portrayed differently, when no picture is attached to it! And lastly, we implemented a footer, that has the same functionality as a header and is helpful, when scrolling through the website! We also adjusted the homepage view for actual smartphone users as you can see in the next image.
  
![HomepageSmartphone](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/39e9e851-3b3d-4b9c-8683-43b3838a748f)
  
#### Functioning Login!
We can finally sign in onto our application. In the image below, you can see the current login dialog where you can enter your username and password. If you leave out at least one of the fields, you'll get an error message below. You can also tick a 'Remember me?'-checkbox, however, this hasn't been fully implemented yet. The username and password input is checked against the database and if everything is fine, you'll be redirected to the profile page of the user (is in progress atm).  
  
![Signin](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/eaf95124-4c35-440c-b17f-f1fee479d981)
  
We are also currently working on the registration process with email verification. After entering valid user data, you'll be asked to enter the verification code that we have sent to your e-mail address like in the images below.
  
![EnterCode](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/de2a824c-f364-4dfc-9903-1ba393a85d4b)
  
![Email](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/9d3f150b-7d67-4d32-b6dd-aac1c1617c67)

### Developments on the backend 🧮
Our progress regarding the connection between the backend and the frontend has increased greatly, since we have launched our backend onto a BWcloud server. Thus, the frontend can receive data which is stored in our actual database. We have also updated our event table as well as the respective endpoints. The changes are in the graph below! Although the Event and Post table seem very similar, we have decided to seperate both from each other to increase performance and speed when retrieving data from these tables (UserID and PictureId are foreign keys, but simplified in the image).

![BlogEventTables](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/526bf187-8e70-4858-aa1a-390d2cd1dba2)
  

### Sprint 8: What have we achieved ❓ What's to come❔
✅ Frontend: Implement footer
✅ Frontend: Adapt login-dialog
✅ Frontend: Implement 'Scroll-Up'-button
✅ Frontend: Implement 'Saved-Posts'-Component
✅ Frontend: Refactor interactions
✅ Frontend: Switch back to BrowserRouter
✅ Frontend: Implement smartphone views for homepage
✅ Backend: Host application on BWCloud-Server
✅ Backend: Setup Email Server
✅ Backend: Refactor Event-Endpoints
✅ Database: Create Event tables, trigger and views
🚧 Frontend: Add PWA shortcuts for common pages
🚧 Frontend: Implement PostDetail-component
🚧 Frontend: Implement essential tests
🚧 Frontend: Verification of email address
🚧 Frontend: Handling of "null"-value images
🚧 Frontend: Implement Comment-component
🚧 Frontend: Create mockups for profile customization, imprint, privacy policy and and tag overview page
🚧 Frontend/Backend: Implement sign in via Google
🚧 Backend: Adapt Friendship Endpoints
🚧 Backend: Adapt SavedPost Endpoints
🚧 Database: Create Logtable for Activities
🚧 Database: Create Concept and Endpoints for a notification system
🚧 Remove Herobrine

Until next week 💯

Your DHBWhub Team 🚀

**Comment by HowDoIUseGit** [comment]
Your progress report is fantastic! It's evident you've put a lot of effort into your testing plan, backend, and frontend development. The addition of new components and functionalities is impressive, and your achievements for Sprint 8 are commendable. Keep up the excellent work! 🚀
Your CookPit

**Comment by Umbranecis** [comment]
This report is phenomenal. You really can see every step that was taken this week and where it should lead in the future. I especially love the list at the end, where you show what exactly is done already and what is not done yet.
Keep up the really good work!
Greeting from Team Plapy
